dNet/TcpSession.cpp

```diff
@@ -0,0 +1,307 @@
+	}
+}
+
+boost::asio::basic_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>& TcpSession::GetSocket() {
+	if (m_UseTls) {
+		return m_SSLStream->lowest_layer();
+	} else {
+		return *m_Socket;
+	}
+}
+
+void TcpSession::StartAsyncRead() {
+	if (!m_IsConnected) {
+		return;
+	}
+
+	m_HeaderBuffer.resize(sizeof(uint32_t));
+
+	auto readHeaderCallback = [this](const boost::system::error_code& ec, std::size_t bytesTransferred) {
+		OnReadHeader(ec, bytesTransferred);
+	};
+
+	try {
+		if (m_UseTls) {
+			boost::asio::async_read(*m_SSLStream, boost::asio::buffer(m_HeaderBuffer), readHeaderCallback);
+		} else {
+			boost::asio::async_read(*m_Socket, boost::asio::buffer(m_HeaderBuffer), readHeaderCallback);
+		}
+	} catch (const std::exception& e) {
+		LOG("Exception: %s", e.what());
+		Close();
+	}
+}
+
+void TcpSession::OnReadHeader(const boost::system::error_code& ec, std::size_t bytesTransferred) {
+	if (!ec) {
+		uint32_t size = 0;
+		std::memcpy(&size, m_HeaderBuffer.data(), sizeof(size));
+
+		m_BodyBuffer.resize(size);
+
+		auto readBodyCallback = [this](const boost::system::error_code& ec, std::size_t bytesTransferred) {
+			OnReadBody(ec, bytesTransferred);
+		};
+
+		if (m_UseTls) {
+			boost::asio::async_read(*m_SSLStream, boost::asio::buffer(m_BodyBuffer), readBodyCallback);
+		} else {
+			boost::asio::async_read(*m_Socket, boost::asio::buffer(m_BodyBuffer), readBodyCallback);
+		}
+	} else {
+		LOG("Read header error: %s", ec.message().c_str());
+		Close();
+	}
+}
+
+void TcpSession::OnReadBody(const boost::system::error_code& ec, std::size_t bytesTransferred) {
+	if (!ec) {
+		auto* packet = new Packet();
+		packet->data = new uint8_t[m_BodyBuffer.size()];
+		std::memcpy(packet->data, m_BodyBuffer.data(), m_BodyBuffer.size());
+		packet->length = m_BodyBuffer.size();
+		packet->bitSize = packet->length * 8;
+		packet->systemAddress = m_SystemAddress;
+
+		if (packet->data[0] == ID_USER_PACKET_ENUM && !m_ReadyForUserPackets) {
+			LOG("Received early user packet from %s:%d", m_SystemAddress.ToString(false), m_SystemAddress.port);
+			m_UserPacketQueue.push(packet);
+		}
+		else if (packet->data[0] == ID_INTERNAL_PING) {
+			HandlePing(packet);
+		}
+		else if (packet->data[0] == ID_CONNECTED_PONG) {
+			HandlePong(packet);
+		}
+		else {
+			m_Server->OnReceive(packet);
+		}
+
+		StartAsyncRead();
+	} else {
+		LOG("Read body error: %s", ec.message().c_str());
+		Close();
+	}
+}
+
+void TcpSession::OnWrite(const boost::system::error_code& ec, std::size_t bytesTransferred) {
+	if (ec) {
+		LOG("Write error: %s", ec.message().c_str());
+		Close();
+	}
+}
+
+void TcpSession::HandlePing(Packet* packet) {
+	RakNet::BitStream bitStream;
+	bitStream.Write<uint8_t>(ID_CONNECTED_PONG);
+	bitStream.Write<uint8_t>(packet->data[1]);
+	bitStream.Write<uint8_t>(packet->data[2]);
+	bitStream.Write<uint8_t>(packet->data[3]);
+	bitStream.Write<uint8_t>(packet->data[4]);
```

Comment:
This could access undefined heap memory if data is < 5 in size, i would consider bounds checks here on a packet